<feColorMatrix> filter doesn't work properly when defined inside then set directly on the <svg> element

tuankiet65 opened the pull request:
#### 0dac14ca10cb771ef13516f2a82581b832f8a1c2
<pre>
&lt;feColorMatrix&gt; filter doesn&apos;t work properly when defined inside then set directly on the &lt;svg&gt; element
<a href="https://bugs.webkit.org/show_bug.cgi?id=225120">https://bugs.webkit.org/show_bug.cgi?id=225120</a>
rdar://77522728

Reviewed by Said Abou-Hallawa.

When a filter is specified on an SVG root, the SVG rendering codepath applies the
filter on the SVG root. Then, the RenderLayer of that SVG applies that filter again,
resulting in the filter being applied twice. Fix this by instructing RenderLayer not
to apply filters when rendering SVG roots.

Tests: LayoutTests/svg/filters/filter-specified-on-svg-root.html

* LayoutTests/svg/filters/filter-specified-on-svg-root-expected.html: Added.
* LayoutTests/svg/filters/filter-specified-on-svg-root.html: Added.
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::paintsWithFilters const):
(WebCore::RenderLayer::calculateClipRects const):

Canonical link: <a href="https://commits.webkit.org/265135@main">https://commits.webkit.org/265135@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/610c8946183ef52dadb47ff219133d5f831294e9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/9843 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/10091 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/10337 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/11495 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/9563 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/9852 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/12075 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/10042 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/12498 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/9997 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/10799 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/8349 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/11895 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/8106 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/8933 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/16291 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/9204 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/9082 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/12371 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/9578 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/7850 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/8743 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/8830 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2382 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/12967 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/9352 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WebCore/rendering/svg/SVGRenderingContext.cpp:
```diff
@@ -161,7 +161,10 @@ void SVGRenderingContext::prepareToRenderSVGContent(RenderElement& renderer, Pai
+
+        // HACK: don't apply filter defined in a SVG root element (which affects the entire SVG document),
+        // since it will be applied later, and we do not want to apply the filter twice.
+        if (m_filter && !m_filter->isIdentity() && !renderer.isSVGRootOrLegacySVGRoot()) {
```
Please do not include "HACK" in the comment. We should strive hard not to add hacks to the code.

And I do not think this is the correct approach. The green faded rectangle is displayed correctly in GitHub without your patch. It seems there is case where only one filter is applied. Please check the test case with your local changes. Either `SVGFilter::apply()` or `CSSFilter::apply()` should be called for this test case.

Also we should check `renderer.isSVGRootOrLegacySVGRoot()` before `m_filter->isIdentity()` since the later is more expensive.

tuankiet65 replied:
I can see the difference on Safari; it's subtle but if you have both the test case and expected SVG on different tabs, switching between them will show the difference.

shallawa replied:
In the Files Changed tab, I do not see a difference. And I debugged this page and I found only `SVGFilter::apply()` is called.

shallawa commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -5567,14 +5567,13 @@ void RenderLayer::clearLayerScrollableArea()
 void RenderLayer::updateFiltersAfterStyleChange(StyleDifference diff, const RenderStyle* oldStyle)
 {
-    if (!hasFilter()) {
+    if (!hasFilter() || (hasFilter() && renderer().isSVGRootOrLegacySVGRoot())) {
```
Ditto.

[Not new] Shouldn't we call `paintsWithFilters()` here instead of just `hasFilter()`?

tuankiet65 replied:
Possibly, but I didn't understand the code enough to make the call. Let me try and report back.

shallawa commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -884,7 +884,7 @@ bool RenderLayer::canRender3DTransforms() const
 bool RenderLayer::paintsWithFilters() const
 {
-    if (!renderer().hasFilter())
+    if (!hasFilter() || (hasFilter() && renderer().isSVGRootOrLegacySVGRoot()))
```
You do not have to check for `hasFilter()` in the second clause:
```
if (!hasFilter() || renderer().isSVGRootOrLegacySVGRoot())
```
This is because `A || !AB <=> A || B`

Or you can add the new statement:
```
if (renderer().isSVGRootOrLegacySVGRoot())
    return false;
```

shallawa commented on LayoutTests/svg/filters/filter-specified-on-svg-root.html:
```diff
@@ -0,0 +1,14 @@
+    </filter>
+    <rect width="100" height="100" fill="green" />
+  </svg><svg width="100" height="100">
+    <rect width="100" height="100" fill="green" fill-opacity="0.6" />
+  </svg>
```
These two lines seem to be left by mistake.

shallawa replied:
Oh I think I was confused by this markup. I did not see the opening <svg> tag.
```
</svg><svg width="100" height="100">
```

tuankiet65 replied:
Yeah, the ending `</svg>` of the first SVG has to be on the same line as the opening `<svg>` of the second SVG, so they'd be next to each other without gaps. Or there might a trick to achieve the same effect.

tuankiet65 committed:
<feColorMatrix> filter doesn't work properly when defined inside then set directly on the <svg> element
https://bugs.webkit.org/show_bug.cgi?id=225120
rdar://77522728

Reviewed by Said Abou-Hallawa.

When a filter is specified on an SVG root, the SVG rendering codepath applies the
filter on the SVG root. Then, the RenderLayer of that SVG applies that filter again,
resulting in the filter being applied twice. Fix this by instructing RenderLayer not
to apply filters when rendering SVG roots.

Tests: LayoutTests/svg/filters/filter-specified-on-svg-root.html

* LayoutTests/svg/filters/filter-specified-on-svg-root-expected.html: Added.
* LayoutTests/svg/filters/filter-specified-on-svg-root.html: Added.
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::paintsWithFilters const):
(WebCore::RenderLayer::calculateClipRects const):

Canonical link: https://commits.webkit.org/265135@main
